### PR TITLE
Implement GetVertexAttrib

### PR DESCRIPTION
**Pull request by dzbarsky**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10776)

### CODE REVIEW

**Comment by dzbarsky** [comment]
@emilio Thoughts on storing the 0'th vertex attrib in WebGLContext vs webrender_traits? The context seems to be where the state is so it makes sense to track there i guess.

Also some tests fail because it looks like VertexAttrib[1234]fv don't end up actually setting while VertexAttrib[1234]f do. Another thing we should track down.

**Comment by dzbarsky** [comment]
@emilio ok should be ready for review :)

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -122,6 +122,8 @@ pub struct WebGLRenderingContext {
     bound_buffer_array: MutNullableHeap<JS<WebGLBuffer>>,
     bound_buffer_element_array: MutNullableHeap<JS<WebGLBuffer>>,
     current_program: MutNullableHeap<JS<WebGLProgram>>,
+    #[ignore_heap_size_of = "Because it's small"]
```
Comment (dzbarsky) [inline components/script/dom/webglrenderingcontext.rs]: this was a hack, we probably need to implement this somewhere right?

**Comment by emilio** [comment]
-S-awaiting-review +S-awaiting-answer 

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/webglrenderingcontext.rs, line 125 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KFriKdq-KpixwrESAB6-r3-125:-KGLpGbMFWjDmuwOCYWI:1314280544) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L125)):*
Yeah, https://github.com/servo/heapsize is the place to do that.

Not a huge deal (it has heapsize zero), but worth having.

---

*[components/script/dom/webglrenderingcontext.rs, line 1001 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpMwBe2Xswy7TyneN:-KGLpMwBe2Xswy7TyneO:-2015213) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1001)):*
Probably not too difficult to check for `constants::VERTEX_ATTRIB_ARRAY_BUFFER_BINDING` and return the current `bound_buffer_array`?

---

*[components/script/dom/webglrenderingcontext.rs, line 1018 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpjvviHM66z95pYl-:-KGLpjvviHM66z95pYl0:-1630889715) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1018)):*
Instead of this ton of panics, it may be worth to have a wildcard like `_ => unreachable!("Invalid parameter type for vertex attribute")`.

This applies to other sites where we use it.

---

*[tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini, line 3 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLqN9qfcYFR9RMfpz3:-KGLqN9qfcYFR9RMfpz4:1131481715) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini#L3)):*
Are this tests still failing? They shouldn't fail, or at least with these messages, any clues on why?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10776#-:-KGLrb5yrZZ05NFGGpKK:453359453)*

**Comment by dzbarsky** [comment]
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/webglrenderingcontext.rs, line 125 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KFriKdq-KpixwrESAB6-r3-125:-KGVPbr3NIV9dI9eHsHf:-1202985255) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L125)):*
are you suggesting we can use VERTEX_ATTRIB_ARRAY_BUFFER_BINDING to avoid storing the 0'th vertex attrib? if so this will go away...

---

*[components/script/dom/webglrenderingcontext.rs, line 1001 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpMwBe2Xswy7TyneN:-KGVPVOqAxYZvKs4LzBE:-1188923373) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1001)):*
sorry, what is VERTEX_ATTRIB_ARRAY_BUFFER_BINDING?

---

*[components/script/dom/webglrenderingcontext.rs, line 1018 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpjvviHM66z95pYl-:-KGVP3UbITfoUwDVZ420:-233470090) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1018)):*
We do this all over the place, let's fix in a followup?

---

*[tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini, line 3 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLqN9qfcYFR9RMfpz3:-KGVOg2NlPhKFN51FGAC:843353120) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini#L3)):*
Yeah, it seems like the VertexAttrib vector methods are all busted and don't actually set things correctly - I debugged a little and it looks like going from JSObject to slice returns None but I haven't had time to figure out why yet

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10776)*

**Comment by emilio** [comment]
Looks good to me given the situation with the failing tests, let me know if you want to implement the `constants::VERTEX_ATTRIB_ARRAY_BUFFER_BINDING` thing (shouldn't be hard), and otherwise this is good to land.

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/webglrenderingcontext.rs, line 1001 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpMwBe2Xswy7TyneN:-KGXUzNErrol4-CQhn-y:2089297300) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1001)):*
No, I'm suggesting to implement [this comment](https://github.com/servo/webrender_traits/blob/master/src/webgl.rs#L646) here, shouldn't be difficult.

---

*[components/script/dom/webglrenderingcontext.rs, line 1018 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLpjvviHM66z95pYl-:-KGXVK4J3CLM8vXVd9N0:1689193178) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/components/script/dom/webglrenderingcontext.rs#L1018)):*
Sounds good.

---

*[tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini, line 3 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10776#-KGLqN9qfcYFR9RMfpz3:-KGXVn8aWyrinv9Vl-2R:1138884762) ([raw file](https://github.com/servo/servo/blob/dbf8915f0fa0fdca739a27e7a4eea41558d00672/tests/wpt/metadata/webgl/conformance-1.0.3/conformance/attribs/gl-vertex-attrib.html.ini#L3)):*
Hum... Yeah I see.

That's because the tests call these functions with arrays, which should use the `sequence<GLFloat>` version of the `vertexAttrib` calls.

That overload is commented right now because it conflicts with the `object` overload, so we don't expect getting them until we get `Float32Array`, etc... support.

That means that our `object` overload gets called, but we can't get a buffer because they're not `Float32Array`s.

We *could* do a check for a JS array and use the `FromJSValConvertible` trait, which is what codegen uses, but I'm not sure it's worth it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10776#-:-KGXWiAg-fp218MdRI9y:-1736615132)*

**Comment by dzbarsky** [comment]
@emilio Let's land this and do VERTEX_ATTRIB_ARRAY_BUFFER_BINDING as a followup

**Comment by nox** [comment]
@bors-servo r=emilio since he said "otherwise this is good to land".

@dzbarsky Please don't forget to do or file the follow-up. :)

**Comment by emilio** [comment]
@bors-servo: retry

 * https://github.com/servo/servo/issues/8815

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -83,6 +83,8 @@ pub struct WebGLRenderingContext {
...
+    current_vertex_attrib_0: Cell<(f32, f32, f32, f32)>,
```
Comment (nox) [inline components/script/dom/webglrenderingcontext.rs]: @dzbarsky Do you remember why it's necessary to keep the value of vertex attrib 0 on the DOM side?

Comment (dzbarsky) [reply components/script/dom/webglrenderingcontext.rs]: When I first implemented this in https://github.com/servo/servo/commit/2bf016ff149da46f8957b959a61d5c5a6afb682a `GetVertexAttrib` had a path to serve 0/CURRENT_VERTEX_ATTRIB from DOM side - I don't recall why I did this though...perhaps it was a fast path or something?

Comment (dzbarsky) [reply components/script/dom/webglrenderingcontext.rs]: Oh I see now that this is the original place we added it :) Most likely I was cargo-culting the pattern of retaining state like `current_program`, though this one was likely not necessary. Oops!

Comment (nox) [reply components/script/dom/webglrenderingcontext.rs]: Seems like calling `glGetVertexAttrib*` with index 0 causes issues on some drivers.

Comment (nox) [reply components/script/dom/webglrenderingcontext.rs]: https://www.khronos.org/webgl/wiki/WebGL_and_OpenGL_Differences#Vertex_Attribute_0

**Pull request merged**